### PR TITLE
test(retrieval): add Nemotron VL checkpoint coverage

### PR DESCRIPTION
**Pull request by yuhezhang-ai**
## Summary

Adds scheduled training and checkpoint-robustness coverage for the existing Llama Nemotron VL 1B retrieval recipe. This is the CI-focused subset retained from #2872 after dropping the optimized model implementation; it does not change production model or checkpoint code.

## Why

The retrieval bi-encoder recipe was not selected by the nightly manifest, and the shared launcher routed checkpoint robustness to the LLM harness. As a result, the existing recipe did not continuously verify its consolidated checkpoint through the retrieval-specific reload path, vanilla Hugging Face query/image inference, or training resume.

## Changes

- Select the existing Nemotron VL retrieval recipe for nightly coverage and give its CI runs public MIRACL data plus bounded training overrides.
- Route retrieval bi-encoder configs to the bi-encoder checkpoint-robustness harness.
- Add config-driven vanilla Hugging Face reload checks for both text-query and image-document embeddings.
- Enable consolidated AutoModel reload and resume-continuity coverage.
- Keep the biencoder resume baseline on the same LR decay schedule used to create the checkpoint.
- Set a config-local `5e-2` resume-loss tolerance for expected BF16 drift between
  the independently initialized baseline and checkpoint-producing run. A
  shared-trajectory redesign is tracked in
  [AMINT-227](https://linear.app/nvidia/issue/AMINT-227/redesign-checkpoint-resume-robustness-around-a-shared-training).
- Document the `distributed.static_graph` and unused-parameter settings required when a DDP launch uses gradient accumulation.

## Validation

- 66 focused checkpoint-harness and config-resolver unit tests passed.
- 40 config-resolver tests passed after correcting the DDP CI overrides.
- Changed Python files pass Ruff check and format validation.
- `tests/ci_tests/scripts/finetune_launcher.sh` passes `bash -n`.
- All 454 MDX files pass the repository syntax validator. Full local `fern check` was unavailable because the Fern CLI is not installed in this workspace.
- [Pipeline 60004309](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/60004309) exposed an unnecessary CI override of the base recipe's static-graph setting. [Pipeline 60015913](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/60015913) then passed training, consolidated AutoModel reload, and vanilla Hugging Face text/image reload, but exposed that the biencoder resume baseline recomputed a longer LR decay than the checkpoint-producing run.
- [Pipeline 60138543](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/60138543) confirmed perfect consolidated AutoModel and vanilla Hugging Face text/image reload parity (`1.000000` cosine). Its resume phase used the corrected LR schedule, but the independently initialized baseline had already diverged before checkpointing; the largest post-checkpoint loss difference was `0.03125`, motivating the config-local `0.05` tolerance.
- Exact-head scoped EOS nightly/checkpoint-robustness rerun: after initial attempts were blocked by the EOS outage, [job 378403572](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/jobs/378403572) passed against `2f38711e1`. The 50-step finetune and biencoder checkpoint-robustness test completed successfully, covering consolidated AutoModel reload, vanilla Hugging Face query/image reload, and training resume.

### CODE REVIEW

**Comment by yuhezhang-ai** [comment]
/ok to test [227e76f](https://github.com/NVIDIA-NeMo/Automodel/pull/3276/commits/227e76f0556e0fb91590b2009decb7c34d67e76b)

File: examples/retrieval/bi_encoder/nemotron_vl_1b/README.md [hunk 1]
```diff
@@ -146,3 +146,16 @@ multi-node training on a Slurm cluster, use `sbatch`.
 ```bash
 torchrun --nproc-per-node=8 examples/retrieval/bi_encoder/finetune.py --config examples/retrieval/bi_encoder/nemotron_vl_1b/nemotron_vl_1b_example.yaml
 ```
+
+The default batch settings use one local batch on each of eight GPUs, so they do not use gradient accumulation and
+`distributed.static_graph: true` is appropriate. If you reduce the GPU count without reducing `global_batch_size`,
```
Comment (rnyak) [inline examples/retrieval/bi_encoder/nemotron_vl_1b/README.md]: maybe rephrase as: "The default batch settings use one local batch on each of eight GPUs, so they do not use gradient accumulation and thus distributed.static_graph: true is appropriate."

File: examples/retrieval/bi_encoder/nemotron_vl_1b/README.md
Comment (jgerh) [inline examples/retrieval/bi_encoder/nemotron_vl_1b/README.md]:
```suggestion
This example shows how to fine-tune the
[nvidia/llama-nemotron-embed-vl-1b-v2](https://huggingface.co/nvidia/llama-nemotron-embed-vl-1b-v2)
model for visual document retrieval. The model embeds document pages supplied
as images, text, or combined image and text inputs. Documents can be retrieved
from a user query in text form. The model supports page images that contain
text, tables, charts, and infographics. Review the model's performance on the
```

Comment (jgerh) [inline examples/retrieval/bi_encoder/nemotron_vl_1b/README.md]:
```suggestion
The `corpus` path points to a directory that contains the corpus in Parquet
format, with a field for the document ID. The `data` key contains the training
samples for contrastive learning: the question, positive samples (`pos_doc`),
and negative samples (`neg_doc`). The positive and negative samples are
document IDs from the corpus and can represent document page images or text
passages.
```

Comment (jgerh) [inline examples/retrieval/bi_encoder/nemotron_vl_1b/README.md]:
```suggestion
to prepare the ColPali example. The notebook performs the following tasks:
```

Comment (jgerh) [inline examples/retrieval/bi_encoder/nemotron_vl_1b/README.md]:
```suggestion
1. Downloads a [training set](https://huggingface.co/datasets/Tevatron/colpali) with mined hard negatives and its
```

Comment (jgerh) [inline examples/retrieval/bi_encoder/nemotron_vl_1b/README.md]:
```suggestion
These files are the source data for both loading methods in this guide. The
notebook does not create normalized Arrow data. That process is a separate CPU
preparation step for full-scale training.
```

Comment (jgerh) [inline examples/retrieval/bi_encoder/nemotron_vl_1b/README.md]:
```suggestion
[MIRACL training set](https://huggingface.co/datasets/nvidia/embed-nemotron-dataset-v1/viewer/MIRACL) for multilingual
```

Comment (jgerh) [inline examples/retrieval/bi_encoder/nemotron_vl_1b/README.md]:
```suggestion
Starting GPU training directly from a large image corpus can leave every
allocated GPU waiting while the corpus loads and its dataset cache builds.
Normalizing on CPU completes that work before GPU allocation.
```

Comment (jgerh) [inline examples/retrieval/bi_encoder/nemotron_vl_1b/README.md]:
```suggestion
For a small verification run or dataset, skip normalization and keep the
original dataset config. The loader reads the notebook's JSON and corpus files
directly:
```

Comment (jgerh) [inline examples/retrieval/bi_encoder/nemotron_vl_1b/README.md]:
```suggestion
The direct path is usually sufficient for text-only retrieval. If startup is
slow, the retrieval data tools also provide a CPU cache-warming script that
keeps this dataset config unchanged.
```

File: examples/retrieval/bi_encoder/nemotron_vl_1b/README.md [hunk 2]
```diff
@@ -146,3 +146,16 @@ multi-node training on a Slurm cluster, use `sbatch`.
 ```bash
 torchrun --nproc-per-node=8 examples/retrieval/bi_encoder/finetune.py --config examples/retrieval/bi_encoder/nemotron_vl_1b/nemotron_vl_1b_example.yaml
 ```
+
+The default batch settings use one local batch on each of eight GPUs, so they do not use gradient accumulation and
+thus `distributed.static_graph: true` is appropriate. If you reduce the GPU count without reducing `global_batch_size`,
+the resulting run uses more than one gradient accumulation step. Update the DDP settings as follows:
```
Comment (jgerh) [inline examples/retrieval/bi_encoder/nemotron_vl_1b/README.md]:
```suggestion
The default batch settings use one local batch on each of eight GPUs, so
gradient accumulation is disabled and `distributed.static_graph: true` is
appropriate. If you reduce the GPU count without reducing `global_batch_size`,
the resulting run uses more than one gradient accumulation step. Update the
distributed data parallel (DDP) settings as follows:
```

File: examples/retrieval/bi_encoder/nemotron_vl_1b/README.md
Comment (jgerh) [inline examples/retrieval/bi_encoder/nemotron_vl_1b/README.md]: Please verify this edit. Removed the placeholder ellipses and closed the neg_doc array and sample object. The original was invalid JSON and could cause parser errors if copied.

```suggestion
        {"id": "5132"}
      ]
    }
```

File: docs/guides/llm/retrieval-finetune.mdx
Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
Run a single-GPU basic verification test first. The timestamped checkpoint directory keeps this command from silently
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
| Component | Bi-Encoder | Cross-Encoder |
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
| Model `config.model_type` | Bi-Encoder Behavior | Cross-Encoder Behavior | Effective Retrieval Keyword Arguments |
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
Hugging Face `AutoModel` fallback can keep the source model's original causal behavior, which might be lower quality for
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
| Data Path | Use When | Notes |
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
| Source | Required Query Field | Required Document Fields |
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
| Inline JSONL | `query` or `question` | Non-empty `pos_doc`, and present `neg_doc` |
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
`neg_doc` must be present for local JSON and JSONL sources. It can be `[]` only when `n_passages: 1`. When
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx:182]:
```suggestion

</Warning>
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
to the cross-encoder values in the following example. Also set `model.num_labels: 1`, set the loss temperature under
`model.temperature`, replace `q_max_len` and `p_max_len` with `rerank_max_length` in the collator, and use a separate
```

File: docs/guides/llm/retrieval-finetune.mdx [hunk 3]
```diff
@@ -461,6 +461,12 @@ gradient_accumulation_steps = global_batch_size / (local_batch_size * data_paral
 For the supported data-parallel FSDP2 configuration, `data_parallel_size` is the total GPU count. For example, two
 8-GPU nodes have `data_parallel_size: 16`.
 
+For DDP recipes, keep `distributed.static_graph: false` whenever `gradient_accumulation_steps > 1`. If the model
```
Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
For Distributed Data Parallel (DDP) recipes, keep `distributed.static_graph: false` whenever
`gradient_accumulation_steps > 1`. If the model
```

File: docs/guides/llm/retrieval-finetune.mdx
Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
held-out retrieval file. Use `RetrievalDatasetConfig` for every supported source. It selects the loading path from the
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
to make the result repeatable. Include the query count, corpus size, qrels source, judged and unjudged handling, exact
or ANN search settings, K values, and baseline checkpoint. State whether you used confidence intervals or significance
tests. At minimum, specify a consolidated bi-encoder checkpoint, corpus and query tables with stable IDs, and qrels
keyed by those IDs. Also specify the query and passage prefixes, maximum lengths, and K values to report.
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
but might not update during a short basic verification test:
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
Use standard output and error as the live per-step signals today. Watch `loss`, `grad_norm`, learning rate, GPU memory,
and step time before scaling to a longer run. On preempted or timed-out jobs, recent buffered JSONL metrics might be
missing even when standard output and error showed them.
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
Each save creates a versioned directory. The following path is an example:
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
With `save_consolidated: true` and full fine-tuning, AutoModel also writes a Hugging Face-compatible model under the
following path:
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
PEFT runs using LoRA save adapter artifacts under the checkpoint `model/` directory instead of the previously described full
consolidated export path. Resume LoRA training from the AutoModel checkpoint directory. If you need a standalone
checkpoint for inference or mining, first produce a Hugging Face-loadable merged encoder with your adapter workflow.
```

Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion
Hugging Face data on each node or use a shared cache. Budget CPU RAM and local disk per rank for corpus-backed datasets.
```

**Review by jgerh — COMMENTED** [review]
Completed tech pubs review of docs/guides/llm/retrieval-finetune.mdx and examples/retrieval/bi_encoder/nemotron_vl_1b/README.md. Provided a few copyedits and suggested text revisions.

**Review by rnyak — APPROVED** [review]

**Review by athitten — APPROVED** [review]
LGTM, thank you @yuhezhang-ai !

**Review by thomasdhc — APPROVED** [review]
